### ISSUE TITLE
Problem creating derived datatypes with MPI_Type_indexed

### ISSUE THREAD

**Issue by tjahns** [pmodels/mpich#5391]
We found an incorrect handling of derived datatypes created with `MPI_Type_indexed`. My colleague created a quite minimal reproducer. Please see the below file for easy to read code that exhibits the problem. He traced the issue down to how `MPI_Type_indexed` was handled, I verified that the issue exists in 3.4 and later releases of MPICH, but not in 3.3.1 or other implementations.

Note: The problem is not related to packing/unpacking at all, that's just a convenient method to demonstrate the issue with only a single rank. The problem was detected for incorrect data from a message that was sent/received with MPI_Isend/MPI_Irecv on distinct ranks.

```
/**
 *
 * This program demonstrates a bug that occurs when using MPI datatypes.
 * The following MPI implementations exhibit this bug:
 * mpich/3.4.1 3.4 and 3.4.2
 *
 * This problem does not occur with the following MPI implementations:
 * intelmpi/2018.4.274 OpenMPI 4.0.5
 *
 * usage:
 *
 *   mpicc mpich_datatype.c && a.out
 *
 * expected output for defective MPICH versions:
 *
 *   pack_buffer[2]:  3 !=  2
 *   pack_buffer[3]: -2 !=  3
 *   pack_buffer[4]:  5 !=  3
 *   pack_buffer[5]:  6 !=  5
 *   pack_buffer[6]:  7 !=  6
 *   pack_buffer[7]: -3 !=  7
 */

#include <stdlib.h>
#include <stdio.h>

#include <mpi.h>

int main(int argc, char **argv) {

  // init mpi
  MPI_Init(NULL, NULL);

  // generate mpi datatype for packing
  MPI_Datatype pack_dt;
  {
    int count = 5;
    int array_of_blocklengths[5] = {2, 2, 1, 3, 3};
    int array_of_displacements[5] = {1, 2, 3, 5, 9};
    MPI_Datatype oldtype = MPI_DOUBLE;
    MPI_Type_indexed(
      count, array_of_blocklengths, array_of_displacements, oldtype, &pack_dt);
    MPI_Type_commit(&pack_dt);
  }
  double inbuf[] = {-1,  1,  2,  3,
                    -2,  5,  6,  7,
                    -3,  9, 10, 11};

  double ref_pack_buffer[] = {1, 2,  2, 3,  3,  5, 6, 7,  9, 10, 11};
  enum {pack_count = sizeof(ref_pack_buffer) / sizeof(ref_pack_buffer[0])};
  double pack_buffer[pack_count];

  // use generated mpi datatype to pack data
  int pos = 0;
  MPI_Pack(
    inbuf, 1, pack_dt,
    pack_buffer, (int)sizeof(pack_buffer), &pos, MPI_COMM_WORLD);

  // check contents of pack_buffer
  int err_flag = 0;
  for (int i = 0, pos = 0; i < pack_count; ++i) {
    double curr_pack_buffer_content;
    MPI_Unpack(
       pack_buffer, (int)sizeof(pack_buffer), &pos,
       &curr_pack_buffer_content, 1, MPI_DOUBLE, MPI_COMM_WORLD);
    if (curr_pack_buffer_content != ref_pack_buffer[i]) {
      err_flag = 1;
      fprintf(stderr, "pack_buffer[%d]: %2.0lf != %2.0lf\n",
              i, curr_pack_buffer_content, ref_pack_buffer[i]);
    }
  }

  // clean up
  MPI_Type_free(&pack_dt);

  // finalise mpi
  MPI_Finalize();

  return (err_flag)?(EXIT_FAILURE):(EXIT_SUCCESS);
}
```

**Comment by hzhou** [comment]
Thanks for the issue. I confirmed the bug. It is due to a bug in how yaksa checks whether the datatype is contiguous. I'll make the fix soon. In the meantime, use configure option `--with-datatype-engine=dataloop` to bypass the issue.

**Comment by tjahns** [comment]
I just noticed that, months after this was closed, MPICH 3.4.3 was released, not including the fix. We are currently distributing a fix at https://gitlab.dkrz.de/dkrz-sw/yaxt/-/blob/master/config/checkpatch/mpich_3.4_yaksa_hindexed.patch that works for the older yaksa used in mpich 3.4.x
